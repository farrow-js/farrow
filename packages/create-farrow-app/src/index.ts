import fs from 'fs'
import path from 'path'
import minimist from 'minimist'
import { prompt } from 'enquirer'
import { green, blue, stripColors, cyan } from 'kolorist'

const argv = minimist(process.argv.slice(2))

const cwd = process.cwd()

// prettier-ignore
const TEMPLATES = [
  blue('farrow-only'),
  green('farrow-vite-react'),
  cyan('farrow-next-react')
]

const renameFiles: Record<string, string> = {
  _gitignore: '.gitignore',
}

type ValidTemplate = {
  isValid: true
  template: string
}

type InvalidTemplate = {
  isValid: false
  isEmpty?: boolean
  message: string
}

type TemplateStatus = ValidTemplate | InvalidTemplate

const validateTemplate = (template: unknown): TemplateStatus => {
  if (typeof template !== 'string') {
    return {
      isValid: false,
      isEmpty: true,
      message: `Expected template to be a string, instead of ${template}`,
    }
  }

  const availableTemplates = TEMPLATES.map(stripColors)

  if (availableTemplates.includes(template)) {
    return {
      isValid: true,
      template,
    }
  }

  return {
    isValid: false,
    message: `Unexpected template: ${template}, the available templates are ${availableTemplates.join(', ')}`,
  }
}

async function init() {
  let targetDir = argv._[0]

  if (!targetDir) {
    const { name } = await prompt<{ name: string }>({
      type: 'input',
      name: 'name',
      message: `Project name:`,
      initial: 'farrow-project',
    })
    targetDir = name
  }

  const root = path.join(cwd, targetDir)
  console.log(`\nScaffolding project in ${root}...`)

  if (!fs.existsSync(root)) {
    fs.mkdirSync(root, { recursive: true })
  } else {
    const existing = fs.readdirSync(root)
    if (existing.length) {
      const { yes } = await prompt<{ yes: boolean }>({
        type: 'confirm',
        name: 'yes',
        initial: 'Y',
        message: `Target directory ${targetDir} is not empty.\n` + `Remove existing files and continue?`,
      })
      if (yes) {
        emptyDir(root)
      } else {
        return
      }
    }
  }

  // determine template
  let template = argv.t || argv.template

  const templateStatus = validateTemplate(template)

  if (!templateStatus.isValid) {
    const message = templateStatus.isEmpty
      ? 'Select a template:'
      : `${template} isn't a valid template. Please choose from below:`

    const { t } = await prompt<{ t: string }>({
      type: 'select',
      name: 't',
      message,
      choices: TEMPLATES,
    })
    template = stripColors(t)
  }

  const templateDir = path.join(__dirname, `../templates/${template}`)

  const write = (file: string, content?: string) => {
    const targetPath = renameFiles[file] ? path.join(root, renameFiles[file]) : path.join(root, file)
    if (content) {
      fs.writeFileSync(targetPath, content)
    } else {
      copy(path.join(templateDir, file), targetPath)
    }
  }

  const files = fs.readdirSync(templateDir).filter((file) => {
    return file !== 'package.json' && !file.includes('node_modules')
  })

  for (const file of files) {
    write(file)
  }

  const pkg = require(path.join(templateDir, `package.json`))

  pkg.name = path
    .basename(root)
    // #2360 ensure package.json name is valid
    .trim()
    .replace(/\s+/g, '-')
    .replace(/^[._]/, '')
    .replace(/[~)('!*]+/g, '-')

  write('package.json', JSON.stringify(pkg, null, 2))

  const pkgManager = /yarn/.test(process.env.npm_execpath ?? '') ? 'yarn' : 'npm'

  console.log(`\nDone. Now run:\n`)
  if (root !== cwd) {
    console.log(`  cd ${path.relative(cwd, root)}`)
  }
  console.log(`  ${pkgManager === 'yarn' ? `yarn` : `npm install`}`)
  console.log(`  ${pkgManager === 'yarn' ? `yarn dev` : `npm run dev`}`)
}

function copy(src: string, dest: string) {
  const stat = fs.statSync(src)
  if (stat.isDirectory()) {
    copyDir(src, dest)
  } else {
    fs.copyFileSync(src, dest)
  }
}

function copyDir(srcDir: string, destDir: string) {
  fs.mkdirSync(destDir, { recursive: true })
  for (const file of fs.readdirSync(srcDir)) {
    const srcFile = path.resolve(srcDir, file)
    const destFile = path.resolve(destDir, file)
    copy(srcFile, destFile)
  }
}

function emptyDir(dir: string) {
  if (!fs.existsSync(dir)) {
    return
  }
  for (const file of fs.readdirSync(dir)) {
    const abs = path.resolve(dir, file)
    // baseline is Node 12 so can't use rmSync :(
    if (fs.lstatSync(abs).isDirectory()) {
      emptyDir(abs)
      fs.rmdirSync(abs)
    } else {
      fs.unlinkSync(abs)
    }
  }
}

init().catch((e) => {
  console.error(e)
})
