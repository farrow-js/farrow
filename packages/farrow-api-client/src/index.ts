export * from './farrow-api-client';