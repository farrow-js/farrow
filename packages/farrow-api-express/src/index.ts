export * from './apiRouter'
export * from './context'
