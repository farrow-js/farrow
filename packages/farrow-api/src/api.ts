import { Type, Prettier, TypeOf, ToSchemaCtor, SchemaCtorInput, JsonType } from 'farrow-schema'
import { createAsyncPipeline, AsyncPipeline, MaybeAsync, useContainer, Container } from 'farrow-pipeline'

export type { JsonType }

export type Typeable<T = unknown> =
  | T
  | {
      [Type]: T
      description?: string
      deprecated?: string
    }

type TypeableContentType<T extends Typeable> = T extends Typeable<infer U> ? U : never

export const getContentType = <T extends Typeable>(typeable: T): TypeableContentType<T> => {
  return typeable[Type] ?? typeable
}

export const getTypeDescription = (typeable: Typeable<any>): string | undefined => {
  return (typeable as any)?.description
}

export const getTypeDeprecated = (typeable: Typeable<any>): string | undefined => {
  return (typeable as any)?.deprecated
}

export type ApiDefinition<Input extends SchemaCtorInput = any, Output extends SchemaCtorInput = any> = {
  /**
   * input schema of api
   */
  input: Typeable<Input>
  /**
   * output schema of api
   */
  output: Typeable<Output>
  /**
   * description of api
   */
  description?: string
  /**
   * depcreated info of api if needed
   */
  deprecated?: string
}

export type TypeOfTypeable<T extends Typeable<SchemaCtorInput>> = Prettier<TypeOf<ToSchemaCtor<TypeableContentType<T>>>>

export type ApiImpl<T extends ApiDefinition> = (
  input: TypeOfTypeable<T['input']>,
) => MaybeAsync<TypeOfTypeable<T['output']>>

export type ApiPipeline<T extends ApiDefinition = ApiDefinition> = AsyncPipeline<
  TypeOfTypeable<T['input']>,
  TypeOfTypeable<T['output']>
>

export type ApiSchema<T extends ApiDefinition = ApiDefinition> = {
  type: 'Api'
  definition: T
}

export type ApiMethods<T extends ApiDefinition = ApiDefinition> = {
  new: () => ApiType<T>
}

export type ApiType<T extends ApiDefinition = ApiDefinition> = ApiImpl<T> &
  ApiSchema<T> &
  ApiPipeline<T> &
  ApiMethods<T>

export const isApi = <T extends ApiDefinition = ApiDefinition>(input: any): input is ApiType<T> => {
  return typeof input === 'function' && input?.type === 'Api'
}

const useContainerSafe = (): Container | undefined => {
  try {
    return useContainer()
    // eslint-disable-next-line no-empty
  } catch (_) {}
}

export function createApi<T extends ApiDefinition>(definition: T, impl?: ApiImpl<T>): ApiType<T> {
  const apiPipeline: ApiPipeline<T> = createAsyncPipeline()

  const apiSchema: ApiSchema<T> = {
    type: 'Api',
    definition,
  }

  const apiImpl: ApiImpl<T> = (input) => {
    const container = useContainerSafe()
    return apiPipeline.run(input, {
      container,
    })
  }

  const apiMethods: ApiMethods<T> = {
    new() {
      return createApi(definition, impl)
    },
  }

  if (impl) {
    apiPipeline.use(impl)
  }

  return Object.assign(apiImpl, apiSchema, apiPipeline, apiMethods)
}

export const Api = createApi

export type ApiEntries = {
  [key: string]: ApiType | ApiEntries
}
