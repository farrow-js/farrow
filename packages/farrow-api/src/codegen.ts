import { FormatFields, FormatType, FormatTypes, isNamedFormatType } from 'farrow-schema/formatter'
import { FormatEntries, FormatResult, FormatApi } from './toJSON'

export const isInlineType = (input: FormatType) => {
  if (isNamedFormatType(input)) {
    return !input.name
  }
  return true
}

const getTypeName = (input: FormatType): string | null => {
  if (isNamedFormatType(input) && input.name) {
    return input.name
  }
  return null
}

const transformComment = (text: string) => {
  return text
    .split('\n')
    .map((item) => item.trim())
    .filter(Boolean)
    .join('\n*\n* ')
}

const applyIndentForEachLine = (text: string, indent: number) => {
  const indentText = ' '.repeat(indent)
  return text
    .split('\n')
    .map((item) => indentText + item)
    .join('\n')
}

const attachComment = (result: string, options: { [key: string]: string | undefined }) => {
  const list = Object.entries(options)
    .map(([key, value]) => {
      return value ? `* @${key} ${transformComment(value.trim())}` : ''
    })
    .filter(Boolean)

  if (list.length === 0) return result

  const comment = `/**\n${list.join('\n')}\n*/\n`

  return comment + result
}

const getTypeNameById = (typeId: number | string): string => {
  return `Type${typeId}`
}

export const getFieldType = (typeId: number, types: FormatTypes, indent = 2): string => {
  const fieldType = types[typeId]

  const typeName = getTypeName(fieldType)

  if (typeName) {
    return typeName
  }

  if (!isInlineType(fieldType)) {
    return getTypeNameById(typeId)
  }

  if (fieldType.type === 'Scalar') {
    return fieldType.valueType
  }

  if (fieldType.type === 'Record') {
    return `Record<string, ${getFieldType(fieldType.valueTypeId, types)}>`
  }

  if (fieldType.type === 'Literal') {
    const literal = typeof fieldType.value === 'string' ? `"${fieldType.value}"` : fieldType.value
    return `${literal}`
  }

  if (fieldType.type === 'Nullable') {
    return `${getFieldType(fieldType.itemTypeId, types)} | null | undefined`
  }

  if (fieldType.type === 'List') {
    return `(${getFieldType(fieldType.itemTypeId, types)})[]`
  }

  if (fieldType.type === 'Union') {
    return fieldType.itemTypes.map((itemType) => getFieldType(itemType.typeId, types)).join(' | ')
  }

  if (fieldType.type === 'Intersect') {
    return fieldType.itemTypes.map((itemType) => getFieldType(itemType.typeId, types)).join(' & ')
  }

  if (fieldType.type === 'Struct') {
    return `{\n${applyIndentForEachLine(getFieldsType(fieldType.fields, types).join(',\n'), indent)}\n${' '.repeat(
      indent - 2,
    )}}`
  }

  if (
    fieldType.type === 'Strict' ||
    fieldType.type === 'NonStrict' ||
    fieldType.type === 'ReadOnly' ||
    fieldType.type === 'ReadOnlyDeep'
  ) {
    return getFieldType(fieldType.itemTypeId, types)
  }

  if (fieldType.type === 'Tuple') {
    return `[${fieldType.itemTypes.map((itemType) => getFieldType(itemType.typeId, types)).join(', ')}]`
  }

  throw new Error(`Unsupported field: ${JSON.stringify(fieldType, null, 2)}`)
}

export const getFieldsType = (fields: FormatFields, types: FormatTypes): string[] => {
  return Object.entries(fields).map(([key, field]) => {
    const fieldType = types[field.typeId]
    let result = ''

    if (fieldType.type === 'Nullable') {
      result = `${key}?: ${getFieldType(field.typeId, types)}`
    } else {
      result = `${key}: ${getFieldType(field.typeId, types)}`
    }

    return attachComment(result, {
      remarks: field.description,
      deprecated: field.deprecated,
    })
  })
}

export type CodegenOptions = {
  /**
   * emit ApiClient or not
   * default: true
   */
  apiClient?: boolean

  /**
   * add ts-nocheck or not
   */
  noCheck?: boolean | string
}

export const PREFIX_COMMENT = `
/**
 * This file was generated by farrow-api
 * Don't modify it manually
*/
`.trim()

export const JSON_TYPE_DECLARATION = `
export type JsonType =
  | number
  | string
  | boolean
  | null
  | undefined
  | JsonType[]
  | { toJSON(): string }
  | { [key: string]: JsonType }
`.trim()

export const codegen = (formatResult: FormatResult, options?: CodegenOptions): string => {
  const exportSet = new Set<string>()

  const handleTypeDeclaration = (formatType: FormatType): string => {
    if (isInlineType(formatType)) {
      return ''
    }

    if (formatType.type === 'Object' || formatType.type === 'Struct') {
      const typeName = formatType.name!
      const fields = getFieldsType(formatType.fields, formatResult.types)

      if (!typeName) {
        throw new Error(`Empty name of Object/Struct, fields: {${Object.keys(formatType.fields)}}`)
      }

      if (exportSet.has(typeName)) {
        throw new Error(`Duplicate Object/Struct type name: ${typeName}`)
      }

      exportSet.add(typeName)

      const source = `
/**
 * @label ${typeName}
*/
export type ${typeName} = {
${applyIndentForEachLine(fields.join(',\n'), 2)}
}
`
      return source.trim()
    }

    if (formatType.type === 'Union') {
      const typeName = formatType.name!
      const expression = formatType.itemTypes.map((itemType) => getFieldType(itemType.typeId, formatResult.types))
      const source = `
/**
 * @label ${typeName}
*/
export type ${typeName} =
${applyIndentForEachLine(expression.map((item) => `| ${item}`).join('\n'), 2)}
      `

      return source.trim()
    }

    if (formatType.type === 'Intersect') {
      const typeName = formatType.name!
      const expression = formatType.itemTypes.map((itemType) => getFieldType(itemType.typeId, formatResult.types))
      const source = `
/**
 * @label ${typeName}
*/
export type ${typeName} =
${applyIndentForEachLine(expression.map((item) => `& ${item}`).join('\n'), 2)}
`

      return source.trim()
    }

    if (formatType.type === 'Tuple') {
      const typeName = formatType.name!
      const expression = formatType.itemTypes.map((itemType) => getFieldType(itemType.typeId, formatResult.types))

      return `
/**
 * @label ${typeName}
*/
export type ${typeName} = [
${applyIndentForEachLine(expression.join(',\n'), 2)}
]
`.trim()
    }

    throw new Error(`Unsupported type of ${JSON.stringify(formatType, null, 2)}`)
  }

  const handleTypeDeclarations = (formatTypes: FormatTypes) => {
    const result = [] as string[]
    for (const key in formatTypes) {
      const formatType = formatTypes[key]
      const formatedType = handleTypeDeclaration(formatType)
      if (formatedType) {
        result.push(formatedType)
      }
    }
    return result
  }

  const typeDeclarations = [JSON_TYPE_DECLARATION, ...handleTypeDeclarations(formatResult.types)]

  const variableDeclarations: string[] = []

  if (options?.apiClient !== false) {
    typeDeclarations.push(
      `
export type ApiClientLoaderInput = {
  path: string[]
  input: JsonType
}
`.trim(),
    )

    typeDeclarations.push(
      `
export interface ApiClientLoaderOptions {
  batch?: boolean
  stream?: boolean
  cache?: boolean
}
`.trim(),
    )

    typeDeclarations.push(
      `
export type ApiClientOptions = {
  loader: (input: ApiClientLoaderInput, options?: ApiClientLoaderOptions) => Promise<JsonType>
}
`.trim(),
    )

    const handleApi = (api: FormatApi, path: string[]) => {
      const inputType = getFieldType(api.input.typeId, formatResult.types)
      const outputType = getFieldType(api.output.typeId, formatResult.types, 4)
      return `
(input: ${inputType}, loaderOptions?: ApiClientLoaderOptions) => {
  return options.loader(
    {
      path: [${path.map((item) => `'${item}'`).join(', ')}],
      input: input as JsonType,
    },
    loaderOptions
  ) as Promise<${outputType}>
}
    `.trim()
    }

    const handleEntries = (entries: FormatEntries, path: string[] = [], indent = 2): string => {
      const fields = [] as string[]

      for (const key in entries.entries) {
        const field = entries.entries[key]

        if (field.type === 'Api') {
          const sourceText = handleApi(field, [...path, key])
          const result = `${key}: ${sourceText}`

          fields.push(attachComment(result, {
            remarks: field.description,
            deprecated: field.deprecated,
            [`param input -`]: field.input.description,
            returns: field.output.description,
          }))
        } else {
          fields.push(`${key}: ${handleEntries(field, [...path, key])}`)
        }
      }

      return `{\n${applyIndentForEachLine(fields.join(',\n'), indent)}\n${' '.repeat(indent - 2)}}`
    }

    variableDeclarations.push(
      `
export const createApiClient = (options: ApiClientOptions) => {
  return ${handleEntries(formatResult.entries, [], 4)}
}
`.trim(),
    )
  }

  let source = [PREFIX_COMMENT, ...typeDeclarations, ...variableDeclarations].join('\n\n')

  if (options?.noCheck) {
    if (typeof options.noCheck === 'string') {
      source = `
// @ts-nocheck ${options.noCheck}
${source}
      `
    } else {
      source = `
// @ts-nocheck
${source}
      `
    }
  }

  return source.trim()
}
