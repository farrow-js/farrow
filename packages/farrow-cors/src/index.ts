import { Response, useReq, useRes, MaybeAsyncResponse } from 'farrow-http'
import type { IncomingMessage } from 'http'
import type { Middleware } from 'farrow-pipeline'

import Cors, { CorsOptions, CorsOptionsDelegate, CorsRequest } from 'cors'
import { promisify } from 'util'

export { CorsOptions, CorsOptionsDelegate, CorsRequest }

export const cors = (
  options?: CorsOptions | CorsOptionsDelegate<IncomingMessage>,
): Middleware<any, MaybeAsyncResponse> => {
  const cors = promisify(
    Cors(
      typeof options === 'function'
        ? options
        : {
            ...options,
            preflightContinue: true,
          },
    ),
  )

  return async (request, next) => {
    const req = useReq()
    const res = useRes()

    try {
      await cors(req, res)
      if (req.method?.toLowerCase() === 'options') {
        if (typeof options === 'object') {
          if (options.preflightContinue) {
            return next(request)
          }
        }
        return Response.status(204).string('')
      }
      return next(request)
    } catch (error: any) {
      return Response.status(500).text(error.message)
    }
  }
}
