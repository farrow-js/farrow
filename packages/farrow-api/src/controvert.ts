import {
  SchemaCtorInput,
  SchemaCtor,
  String,
  ID,
  Number,
  Int,
  Float,
  Boolean,
  Unknown,
  Any,
  Json,
  ObjectType,
  FieldDescriptor,
  Union,
  Intersect,
  Struct,
  FieldDescriptors,
  Record,
  List,
  Tuple,
  Literal,
  Nullable,
  Strict,
  NonStrict,
  ReadOnly,
  ReadOnlyDeep,
} from 'farrow-schema'
import { Api, ApiEntries, ApiType } from './api'
import type {
  FormatTypes,
  FormatType,
  FormatScalarType,
  FormatObjectType,
  FormatFields,
  FormatField,
  FormatUnionType,
  FormatIntersectType,
  FormatStructType,
  FormatRecordType,
  FormatListType,
  FormatTupleType,
  FormatLiteralType,
  FormatNullableType,
  FormatStrictType,
  FormatNonStrictType,
  FormatReadOnlyType,
  FormatReadonlyDeepType,
} from 'farrow-schema/formatter'
import type { FormatResult, FormatEntries, FormatApi } from './toJSON'

export const controvertEntries = (input: FormatResult): ApiEntries => {
  const types = controvertTypes(input.types)

  const findType = (typeId: number): SchemaCtor => {
    const schemaCtor = types.get(typeId.toString())

    if (!schemaCtor) {
      throw new Error(`Unknown typeId: ${typeId}`)
    }

    return schemaCtor
  }

  const controvertFieldType = (input: FormatField): FieldDescriptor => {
    return {
      __type: findType(input.typeId),
      description: input.description,
      deprecated: input.deprecated,
    }
  }

  const controvertApi = (input: FormatApi): ApiType => {
    return Api({
      input: controvertFieldType(input.input),
      output: controvertFieldType(input.output),
      description: input.description,
      deprecated: input.deprecated,
    })
  }

  const controvertEntries = (input: FormatEntries): ApiEntries => {
    const entries: ApiEntries = {}

    for (const name in input.entries) {
      const item = input.entries[name as keyof typeof input.entries]
      if (item.type === 'Api') {
        entries[name] = controvertApi(item)
      } else {
        entries[name] = controvertEntries(item)
      }
    }

    return entries
  }

  return controvertEntries(input.entries)
}

export const controvertTypes = (input: FormatTypes): Map<string, SchemaCtor> => {
  const controvertType = (input: FormatType): SchemaCtor => {
    switch (input.type) {
      case 'Scalar': {
        return controvertScalarType(input)
      }
      case 'Object': {
        return controvertObjectType(input)
      }
      case 'Union': {
        return controvertUnionType(input)
      }
      case 'Intersect': {
        return controvertIntersectType(input)
      }
      case 'Struct': {
        return controvertStructType(input)
      }
      case 'Record': {
        return controvertRecordType(input)
      }
      case 'List': {
        return controvertListType(input)
      }
      case 'Tuple': {
        return controvertTupleType(input)
      }
      case 'Literal': {
        return controvertLiteralType(input)
      }
      case 'Nullable': {
        return controvertNullableType(input)
      }
      case 'Strict': {
        return controvertStrictType(input)
      }
      case 'NonStrict': {
        return controvertNonStrictType(input)
      }
      case 'ReadOnly': {
        return controvertReadOnlyType(input)
      }
      case 'ReadOnlyDeep': {
        return controvertReadOnlyDeepType(input)
      }
      // for eslint
      default: {
        throw new Error(`Unknown format type: ${input}`)
      }
    }
  }

  const controvertScalarType = (input: FormatScalarType): SchemaCtor => {
    switch (input.valueName) {
      case 'String':
        return String
      case 'ID':
        return ID
      case 'Number':
        return Number
      case 'Int':
        return Int
      case 'Float':
        return Float
      case 'Boolean':
        return Boolean
      case 'Date':
        return Date
      case 'Unknown':
        return Unknown
      case 'Any':
        return Any
      case 'Json':
        return Json
      default:
        throw new Error(`Unknown Scalar Type name: ${input.valueName}`)
    }
  }

  const controvertObjectType = (input: FormatObjectType): SchemaCtor => {
    const fields = controvertFieldsType(input.fields)

    return class Obj extends ObjectType {
      constructor() {
        super()

        for (const name in fields) {
          // @ts-ignore
          this[name] = fields[name]
        }
      }
    }
  }

  const controvertFieldsType = (input: FormatFields): FieldDescriptors => {
    const fields: FieldDescriptors = {}

    for (const name in input) {
      fields[name] = controvertFieldType(input[name])
    }

    return fields
  }

  const controvertFieldType = (input: FormatField): FieldDescriptor => {
    return {
      __type: findType(input.typeId),
      description: input.description,
      deprecated: input.deprecated,
    }
  }

  const controvertUnionType = (input: FormatUnionType): SchemaCtor => {
    const items: SchemaCtorInput[] = input.itemTypes.map(({ typeId }) => findType(typeId))
    return Union(...items)
  }

  const controvertIntersectType = (input: FormatIntersectType): SchemaCtor => {
    const items: SchemaCtorInput[] = input.itemTypes.map(({ typeId }) => findType(typeId))
    return Intersect(...items)
  }

  const controvertStructType = (input: FormatStructType): SchemaCtor => {
    const fields = controvertFieldsType(input.fields)
    return Struct(fields)
  }

  const controvertRecordType = (input: FormatRecordType): SchemaCtor => {
    const item = findType(input.valueTypeId)
    return Record(item)
  }

  const controvertListType = (input: FormatListType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return List(item)
  }

  const controvertTupleType = (input: FormatTupleType): SchemaCtor => {
    const items: SchemaCtorInput[] = input.itemTypes.map(({ typeId }) => findType(typeId))
    return Tuple(...items)
  }

  const controvertLiteralType = (input: FormatLiteralType): SchemaCtor => {
    return Literal(input.value)
  }

  const controvertNullableType = (input: FormatNullableType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return Nullable(item)
  }

  const controvertStrictType = (input: FormatStrictType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return Strict(item)
  }

  const controvertNonStrictType = (input: FormatNonStrictType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return NonStrict(item)
  }

  const controvertReadOnlyType = (input: FormatReadOnlyType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return ReadOnly(item)
  }

  const controvertReadOnlyDeepType = (input: FormatReadonlyDeepType): SchemaCtor => {
    const item = findType(input.itemTypeId)
    return ReadOnlyDeep(item)
  }

  const findType = (typeId: number): SchemaCtor => {
    const schemaCtor = types.get(typeId.toString())

    if (!schemaCtor) {
      throw new Error(`Unknown typeId: ${typeId}`)
    }

    return schemaCtor
  }

  const types = new Map<string, SchemaCtor>()

  for (const id in input) {
    types.set(id, controvertType(input[id]))
  }

  return types
}
