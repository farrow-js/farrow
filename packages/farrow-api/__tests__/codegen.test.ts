import { Api } from '../src/api'
import { toJSON } from '../src/toJSON'
import { codegen } from '../src/codegen'
import {
  Any,
  Float,
  ID,
  Int,
  Intersect,
  Json,
  List,
  Literal,
  Nullable,
  ObjectType,
  Record,
  Struct,
  Type,
  Union,
  Unknown,
  Date,
  Tuple,
  Optional,
} from 'farrow-schema'

const NamedStruct = Struct({
  named: String,
  nest: {
    a: Int,
    nest: {
      b: Float,
    },
  },
})

NamedStruct.displayName = 'NamedStruct'

const NamedUnion = Union(Int, String, Float, {
  named: String,
  nest: {
    a: Int,
    nest: {
      b: Float,
    },
  },
})

NamedUnion.displayName = 'NamedUnion'

const NamedIntersect = Intersect(
  {
    a: Int,
  },
  {
    b: Float,
  },
  {
    c: Number,
  },
  {
    named: String,
    nest: {
      a: Int,
      nest: {
        b: Float,
      },
    },
  },
)

NamedIntersect.displayName = 'NamedIntersect'

const NamedTuple = Tuple(
  { a: Int },
  { b: Float },
  { c: Number },
  {
    named: String,
    nest: {
      a: Int,
      nest: {
        b: Float,
      },
    },
  },
)

NamedTuple.displayName = 'NamedTuple'


class Collection extends ObjectType {
  namedStruct = NamedStruct
  namedUnion = NamedUnion
  namedIntersect = NamedIntersect
  namedTuple = NamedTuple
  number = Number
  int = Int
  float = Float
  string = String
  boolean = Boolean
  id = ID
  date = Date
  nest = Nullable(Collection)
  optional?= Optional(String)
  list = List(Collection)
  struct = Struct({
    named: String,
    nest: {
      a: Int,
      nest: {
        b: Float,
      },
    },
  })
  union = Union(Int, String, Boolean)
  intersect = Intersect(
    {
      a: Int,
    },
    {
      b: Float,
    },
    {
      c: Number,
    },
  )
  any = Any
  unknown = Unknown
  json = Json
  literal = Union(Literal(1), Literal('1'), Literal(false), Literal(null))
  record = Record(Collection)

  describable = {
    description: 'test description',
    deprecated: 'test deprecated',
    [Type]: Int,
  }
}


class A extends ObjectType {
  a = String
}

A.namespace = 'TestNamespace'

class B extends ObjectType {
  b = String
}

B.namespace = 'TestNamespace'

const NamedRecord = Record(Collection)

NamedRecord.displayName = 'NamedRecord'

const methodA = Api(
  {
    input: Struct({
      named: String,
      testNamespace: {
        a: A,
        b: B,
        namedRecord: NamedRecord,
      },
      nest: {
        a: Int,
        nest: {
          b: Float,
        },
      },
    }),
    output: Collection,
  },
  () => {
    throw new Error('No Implementation')
  },
)

const methodB = Api(
  {
    input: Collection,
    output: Struct({
      named: String,
      nest: {
        a: Int,
        nest: {
          b: Float,
        },
      },
    }),
  },
  () => {
    throw new Error('No Implementation')
  },
)

const entries = {
  methodA,
  methodB,
  nest: {
    methodA,
    nest: {
      methodB,
    },
  },
}

describe('generateApi', () => {
  it('can disable emitting api-client', async () => {
    const formatResult = toJSON(entries)

    const source = codegen(formatResult, {})

    expect(source).toMatchSnapshot()
  })

  it('can attach header and footer when codegen', async () => {
    const formatResult = toJSON(entries)

    const source = codegen(formatResult, {
      header: `// header`,
      footer: `// footer`,
    })

    expect(source.includes('// header')).toBe(true)
    expect(source.includes('// footer')).toBe(true)
  })
})
