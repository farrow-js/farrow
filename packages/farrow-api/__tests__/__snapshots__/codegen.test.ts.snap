// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateApi can disable emitting api-client 1`] = `
"/**
 * This file was generated by farrow-api
 * Don't modify it manually
*/

export type JsonType =
  | number
  | string
  | boolean
  | null
  | undefined
  | JsonType[]
  | { toJSON(): string }
  | { [key: string]: JsonType }

/**
 * @label Collection
*/
export type Collection = {
  namedStruct: NamedStruct,
  namedUnion: NamedUnion,
  namedIntersect: NamedIntersect,
  namedTuple: NamedTuple,
  number: number,
  int: number,
  float: number,
  string: string,
  boolean: boolean,
  id: string,
  date: string,
  nest: Collection | null,
  optional?: string | undefined,
  list: (Collection)[],
  struct: {
    named: string,
    nest: {
      a: number,
      nest: {
        b: number
      }
    }
  },
  union: number | string | boolean,
  intersect: {
    a: number
  } & {
    b: number
  } & {
    c: number
  },
  any: any,
  unknown: unknown,
  json: JsonType,
  literal: 1 | "1" | false | null,
  record: Record<string, Collection>,
  /**
  * @remarks test description
  * @deprecated test deprecated
  */
  describable: number
}

/**
 * @label NamedStruct
*/
export type NamedStruct = {
  named: string,
  nest: {
    a: number,
    nest: {
      b: number
    }
  }
}

/**
 * @label NamedUnion
*/
export type NamedUnion =
  | number
  | string
  | number
  | {
    named: string,
    nest: {
      a: number,
      nest: {
        b: number
      }
    }
  }

/**
 * @label NamedIntersect
*/
export type NamedIntersect =
  & {
    a: number
  }
  & {
    b: number
  }
  & {
    c: number
  }
  & {
    named: string,
    nest: {
      a: number,
      nest: {
        b: number
      }
    }
  }

/**
 * @label NamedTuple
*/
export type NamedTuple = [
  {
    a: number
  },
  {
    b: number
  },
  {
    c: number
  },
  {
    named: string,
    nest: {
      a: number,
      nest: {
        b: number
      }
    }
  }
]

export namespace TestNamespace {
  /**
   * @label A
  */
  export type A = {
    a: string
  }
  
  /**
   * @label B
  */
  export type B = {
    b: string
  }
}

export type ApiClientLoaderInput = {
  path: string[]
  input: JsonType
}

declare global {
  interface ApiClientLoaderOptions {}
}

export type ApiClientOptions = {
  loader: (input: ApiClientLoaderInput, options?: ApiClientLoaderOptions) => Promise<JsonType>
}

export const createApiClient = (options: ApiClientOptions) => {
  return {
    methodA: (input: {
      named: string,
      testNamespace: {
        a: TestNamespace.A,
        b: TestNamespace.B,
        namedRecord: Record<string, Collection>
      },
      nest: {
        a: number,
        nest: {
          b: number
        }
      }
    }, loaderOptions?: ApiClientLoaderOptions) => {
      return options.loader(
        {
          path: ['methodA'],
          input: input as JsonType,
        },
        loaderOptions
      ) as Promise<Collection>
    },
    methodB: (input: Collection, loaderOptions?: ApiClientLoaderOptions) => {
      return options.loader(
        {
          path: ['methodB'],
          input: input as JsonType,
        },
        loaderOptions
      ) as Promise<{
        named: string,
        nest: {
          a: number,
          nest: {
            b: number
          }
        }
      }>
    },
    nest: {
      methodA: (input: {
        named: string,
        testNamespace: {
          a: TestNamespace.A,
          b: TestNamespace.B,
          namedRecord: Record<string, Collection>
        },
        nest: {
          a: number,
          nest: {
            b: number
          }
        }
      }, loaderOptions?: ApiClientLoaderOptions) => {
        return options.loader(
          {
            path: ['nest', 'methodA'],
            input: input as JsonType,
          },
          loaderOptions
        ) as Promise<Collection>
      },
      nest: {
        methodB: (input: Collection, loaderOptions?: ApiClientLoaderOptions) => {
          return options.loader(
            {
              path: ['nest', 'nest', 'methodB'],
              input: input as JsonType,
            },
            loaderOptions
          ) as Promise<{
            named: string,
            nest: {
              a: number,
              nest: {
                b: number
              }
            }
          }>
        }
      }
    }
  }
}"
`;
